third-party-synthetic/aws-cloudwatch/dynatrace-canary-exporter.js

```diff
@@ -0,0 +1,433 @@
+            [http, https].forEach((package) => ['get', 'request'].forEach((funcName) => {
+                saveFunc(package, funcName);
+                const originalFunc = package[funcName];
+                package[funcName] = function (urlOrOptions) {
+                    const request = originalFunc.apply(package, arguments);
+                    handleCanaryRequest(request, urlOrOptions, (stepResult) => _stepResults$.push(stepResult));
+                    return request;
+                }
+            }));
+            // ----- Web api Canary instrumentation: END -----
+        }
+
+        function afterCanary() {
+            // Restore the overriden methods
+            restoreSavedFuncs();
+        }
+
+        // ----- Web page Canary step builder: START -----
+        function initDtTestResultHandlers(page, onStepResult) {
+            const responses = [];
+
+            // Save every response so that we can look up status codes
+            page.on('response', async (response) => {
+                responses.push({
+                    url: response.url(),
+                    status: response.status(),
+                });
+            });
+
+            // Each time a page/dom loads we save the result as a step
+            page.on('domcontentloaded', async () => {
+                const metrics = (await page._client.send('Performance.getMetrics')).metrics;
+
+                const metric = metrics.find(m => m.name === 'DomContentLoaded').value;
+                const response = responses.find((response) => page.url() === response.url);
+
+                if (!response) {
+                    if (page.url() !== 'about:blank') {
+                        log.error('DT: A response was unexpectedly missing for the URL: ' + page.url());
+                    }
+                } else {
+                    const status = response.status;
+                    const success = isSuccessfulStatusCode(status);
+                    const title = (await page.title()) || page.url();
+
+                    const stepResult = {
+                        title: title,
+                        startTimestamp: Date.now() - metric,
+                        responseTimeMillis: metric,
+                        errorWrapper: success ? {} : {
+                            error: {
+                                message: `Failed to load: '${title}' (${page.url()}).`,
+                                code: status,
+                            }
+                        }
+                    };
+
+                    onStepResult(stepResult);
+                    log.info(`DT: Step result (web page): ${JSON.stringify(stepResult)}.`);
+                }
+            });
+        }
+        // ----- Web page Canary step builder: END -----
+
+        // ----- Web api Canary step builder: START -----
+        const ignoredApiUrlPatterns = [
+            /^monitoring\..*\.amazonaws.com(\/.*)?$/, // Where synthetics posts canary results

```

Comment:
This is necessary when the web page and web api scripts are merged, because the `synthetics` pacakge seems to use the `https` pacakge during the canary run.